File: frontend/src/format.js
// Money / number formatting shared across the dashboard.

const SYMBOL = { TWD: "NT$", USD: "US$" };

export function money(value, currency = "TWD", digits = 0) {
  if (value == null || Number.isNaN(value)) return "—";
  const sym = SYMBOL[currency] || "";
  const n = Number(value).toLocaleString("en-US", {
    minimumFractionDigits: digits,
    maximumFractionDigits: digits,
  });
  return `${sym}${n}`;
}

export function signedMoney(value, currency = "TWD", digits = 0) {
  if (value == null || Number.isNaN(value)) return "—";
  const sign = value > 0 ? "+" : value < 0 ? "−" : "";
  return `${sign}${money(Math.abs(value), currency, digits)}`;
}

// Compact form for chart scales: NT$8.1M, NT$505K. A chart's value gutter has
// to stay narrow or it crowds the plot on a phone — full-precision figures
// belong in the tooltip and the cards, not on an axis.
export function compactMoney(value, currency = "TWD") {
  if (value == null || Number.isNaN(value)) return "—";
  const sym = SYMBOL[currency] || "";
  const sign = value < 0 ? "−" : "";
  const n = Math.abs(value);
  const [div, suffix] =
    n >= 1e9 ? [1e9, "B"] : n >= 1e6 ? [1e6, "M"] : n >= 1e3 ? [1e3, "K"] : [1, ""];
  const scaled = n / div;
  return `${sign}${sym}${scaled.toFixed(suffix && scaled < 10 ? 1 : 0)}${suffix}`;
}

export function pct(value, digits = 2) {
  if (value == null || Number.isNaN(value)) return "—";
  const sign = value > 0 ? "+" : value < 0 ? "−" : "";
  return `${sign}${Math.abs(value).toFixed(digits)}%`;
}

export function shares(value) {
  if (value == null) return "—";
  return Number(value).toLocaleString("en-US", { maximumFractionDigits: 4 });
}

// Color class for a P/L value (green up / red down / muted flat).
export function plClass(value) {
  if (value == null || value === 0) return "muted";
  return value > 0 ? "up" : "down";
}

export function prettyDate(iso) {
  if (!iso) return "—";
  const d = new Date(iso.slice(0, 10) + "T00:00:00Z");
  if (Number.isNaN(d.getTime())) return iso;
  return d.toLocaleDateString("en-US", {
    year: "numeric",
    month: "short",
    day: "numeric",
    timeZone: "UTC",
  });
}

File: ios/StockTracker/Util/Format.swift
import Foundation

/// Formatting helpers mirroring the web app's format.ts so figures read the
/// same across platforms (NT$ / $, signed percentages, em-dash for nil).
enum Fmt {
    static func money(_ value: Double?, currency: String, digits: Int = 2) -> String {
        guard let v = value, !v.isNaN else { return "—" }
        let symbol = currency == "TWD" ? "NT$" : currency == "USD" ? "$" : ""
        let sign = v < 0 ? "-" : ""
        return "\(sign)\(symbol)\(number(abs(v), digits: digits))"
    }

    static func number(_ value: Double?, digits: Int = 2) -> String {
        guard let v = value, !v.isNaN else { return "—" }
        let f = NumberFormatter()
        f.numberStyle = .decimal
        f.minimumFractionDigits = digits
        f.maximumFractionDigits = digits
        return f.string(from: NSNumber(value: v)) ?? "—"
    }

    /// Compact share count: integers show no decimals, fractional shares keep them.
    static func shares(_ value: Double) -> String {
        if value == value.rounded() { return number(value, digits: 0) }
        return number(value, digits: 4)
    }

    static func pct(_ value: Double?) -> String {
        guard let v = value, !v.isNaN else { return "—" }
        let sign = v > 0 ? "+" : ""
        return "\(sign)\(String(format: "%.2f", v))%"
    }

    static func signedMoney(_ value: Double?, currency: String, digits: Int = 2) -> String {
        guard let v = value, !v.isNaN else { return "—" }
        let sign = v > 0 ? "+" : ""
        return "\(sign)\(money(v, currency: currency, digits: digits))"
    }

    /// Big "net worth" style number — thousands separators, no decimals.
    static func bigMoney(_ value: Double?, currency: String) -> String {
        money(value, currency: currency, digits: 0)
    }

    /// "Mar 4, 2025" from an ISO yyyy-MM-dd (or full timestamp) string.
    static func prettyDate(_ iso: String?) -> String {
        guard let iso, !iso.isEmpty else { return "—" }
        let datePart = String(iso.prefix(10))
        let inFmt = DateFormatter()
        inFmt.dateFormat = "yyyy-MM-dd"
        inFmt.timeZone = TimeZone(identifier: "UTC")
        guard let date = inFmt.date(from: datePart) else { return datePart }
        let out = DateFormatter()
        out.dateFormat = "MMM d, yyyy"
        return out.string(from: date)
    }

    /// Abbreviate large counts (market cap, volume): 1.2B, 340M, 12K.
    static func compact(_ value: Double?) -> String {
        guard let v = value, !v.isNaN else { return "—" }
        let abs = Swift.abs(v)
        let sign = v < 0 ? "-" : ""
        switch abs {
        case 1e12...: return "\(sign)\(String(format: "%.2f", abs / 1e12))T"
        case 1e9...: return "\(sign)\(String(format: "%.2f", abs / 1e9))B"
        case 1e6...: return "\(sign)\(String(format: "%.2f", abs / 1e6))M"
        case 1e3...: return "\(sign)\(String(format: "%.1f", abs / 1e3))K"
        default: return number(v, digits: 0)
        }
    }

    /// Chart time-axis label format matched to the visible span: "Jun 5" for
    /// weeks–months, "Jun" for about a year, "2025" beyond that. A fixed
    /// month-only format repeats the same label on short ranges and drops the
    /// year on long ones.
    static func axisFormat(from first: Date, to last: Date) -> Date.FormatStyle {
        let days = last.timeIntervalSince(first) / 86_400
        if days <= 120 { return .dateTime.month(.abbreviated).day() }
        if days <= 550 {
            // Month-only labels turn ambiguous once the span straddles New
            // Year (e.g. the value chart's MAX from Jan 2025): "Mar" could be
            // either year, so append it.
            let crossesYear = Calendar.current.component(.year, from: first)
                != Calendar.current.component(.year, from: last)
            return crossesYear ? .dateTime.month(.abbreviated).year()
                               : .dateTime.month(.abbreviated)
        }
        return .dateTime.year()
    }
}
